만국박람회 [STEP 2] liam.777, joel.inju

liam-777 opened the pull request:
# STEP2
@dacodaco  
STEP2 PR 리뷰 요청드립니다 😀

# 구현한 것

* MVC 패턴으로 구현

* Common
  * `Constant` : 앱에서 필요한 상수들을 static으로 정의

* Views
  * 스토리보드, nib파일, cell
* Controllers
  * `ExpositionViewController`
    * 초기 만국박람회 소개 화면의 ViewController
    * prepare 메소드를 오버라이드 하여
      - KoreanEntryListViewController의 dataSource, delegate에 의존성 주입
  * `KorenEntryListViewController`
    * 한국의 출품작 리스트를 테이블 뷰를 이용해서 보여주는 화면
  * `KorenEntryDetailViewController`
    * 한국 출품작 상세 화면의 ViewController
  * `KorenEntryListTableViewDelegate`
    * NSObject를 상속하고 UITableViewDelegate를 채택한 클래스
    * `KoreanEntryListViewController` 의 tableview에서 사용하게될 delegate 정의
    * `didSelectRowAt` 에서 `NotificationCenter` 를 이용해서 선택된 `indexpath.row`를 옵저버에게 전달합니다.
* Models
  * `KoreanEntry` : 한국 출품작 구조체
  * `UniversalExposition` : 만국박람회 구조체
  * `ExpositionApiMock` 
    * 임시로 ExpositionApu 역할을 수행할 Mock 클래스
    * `ExpositionApi` 프로토콜을 채택하여 필요한 메서드를 구현
    * `ApiMock` 클래스를 상속받아 Mock 클래스로서의 역할 수행
  * `ApiMock` 
    * 임시로 API 역할을 수행하는 클래스
    * JSON 파일을 파싱하여 completion handler를 통해서 전달
    * Generic을 이용하여 json 형식에 따라 여러 타입의 구조체로 반환하도록 구현
  * `ExpositionApi`
    * 만국박람회에 필요한 Api를 선언한 프로토콜
    * 구조체 KoreanEntry과 UniversalExposition을 fetch 해오는 메소드를 가지고 있는 프로토콜
  * `ExpositionApiInjection`
    * ExpositionApi 프로토콜로 선언된 api 프로퍼티에  ExpositionApiMock 클래스를 주입해주는 역할을 하는 클래스
    * static method를 사용하여 구현
  * `KorenEntryListTableViewDataSource`
    * `KoreanEntryListViewController` 의 tableview에서 사용하게될 datasource 정의
    * NSObject를 상속하고 UITableViewDataSource를 채택한 클래스

# 궁금한 점

1. ExpositionApiInjection 클래스 (싱글턴에 대한 고찰)

   * dataSource, delegate 관련

     * 야곰 라이브세션 때 테이블 뷰의 dataSource와 delegate을 ViewController에서 채택하여 사용하는 것이 아니라 별도의 클래스로 분리하여 구현하는 것에 대한 소개를 받았습니다.

     * 저희도 이번에 delegate, dataSource를 `KorenEntryListViewController`로부터 분리하고자 노력을 했고 자연스럽게 의존성 주입에 대한 고민이 이어졌습니다.

     * API 클래스에 대해서는 싱글톤으로 설계하는 것이 가장 편한 방법으로 보이지만, 싱글턴을 사용하면 코드의 결합도가 높아진다고 생각해서 지양하도록 하였습니다.

   1. **그래서** ExpositionApiInjection 클래스를 만들어서 static method로 ExpositionApiMock 클래스를 주입해주는 기능을 만들고
   2. 뷰 컨트롤러에서 ExpositionApi 프로토콜로 선언된 api 프로퍼티에 주입해주는 코드를 만들었습니다.
   3. api 프로퍼티는 ExpositionApi 프로토콜로 선언했고, 뷰컨트롤러에서 ExpositionApiMock 클래스를 직접 주입해주는 코드보다 캡슐화 관점에서 좋을 것 같고 (api 프로퍼티가 데이터를 서버에 받아오는지, 내부 디비에서 가져오는지, 더미데이터에서 가져오는지 뷰 컨트롤러에서는 모르게 하기 위함 -> **단일책임을 위해서는 이 방법이 옳은 것 같지만, 가독성 측면에서는 어떨지 잘 모르겠습니다.** ) 
   4. 이 구현에 대한 코다의 생각이 궁금합니다! 더 좋은 방식 혹은, 저희가 잘못 생각하였거나 생각하지 못한 부분이 있으면 알고 싶습니다!

2. 저희 둘다 MVC 패턴이 아닌 다른 아키텍처 패턴으로 공부를 시작해서, 현재 프로젝트의 구조가 MVC 패턴에 충실한 구조인지 잘 모르겠습니다. 이에 대한 코다는 어떻게 생각하는지 궁굼합니다!

3. Delegate를 클래스로 분리할 경우 NotificationCenter를 이용한 방법

   1. UITableViewDelegate를 뷰컨트롤러에서 채택하지 않고, 커스텀 클래스를 만들어서 UITableViewDelegate을 채택하도록 구현했습니다
   2. 이렇게 커스텀 클래스를 만들고 나니,
      1. tablewView-didSelectRowAt 메소드에서 prepare 메소드를 사용할 수 없었고(뷰컨트롤러가 아니기 때문에)
      2. prepare를 뷰컨트롤러에서 사용해야 해서 뷰컨트롤러에 해당 셀의 indexPath.row 정보를 넘겨주어야 했고
      3. 따라서 NotificationCenter를 이용하는 방법으로 구현하였습니다
      4. 커스텀 클래스를 만들어서 UITableViewDelegate을 채택하여 구현하게 되면, 이 방법이 최선인지 궁금합니다!

4. 테이블 뷰(koreanEntryListTableView)에서 아이템을 클릭해서 상세 보기 화면으로 넘어갈 때 

   1. Expo1900[12841:6462609] [Storyboard] Unknown class Ko in Interface Builder file. 로그가 찍히는데
   2. 아직 이 버그를 명확하게 해결할 방법을 찾지 못했습니다. 코다가 이에 대해 알고 있으시다면 공유해주시면 좋을 것 같습니다!

5. custom delegate class는 controllers 폴더에, custom datasource class 는 model 폴더에 분리해서 넣었는데 괜찮을지?

   1. 파일을 폴더로 분리하는 과정에서 위와 같이 분리하였습니다.
   2. 각각의 역할이 컨트롤러와 모델과 가깝다고 생각하여 저렇게 하였는데, 코다는 어떻게 생각하시는지 궁금합니다.

6. UINib파일을 이용하여 테이블 뷰 셀을 구성하였을 때의 장점 & 단점?

   1. 저희는 UINib파일을 사용했습니다.
   2. UINib파일을 사용하면 인터페이스 빌더에서 가시성이 높아져서 이렇게 사용했는데
   3. 코다가 생각하는 UINib파일을 이용해 구현하게 되었을 시 장점과 단점이 궁금합니다!

7. api를 이용해서 data를 fetch해오는 코드를 라이프사이클의 어디에서 사용할 지

   1. WillAppear? DidLoad? 혹은 다른 라이프 사이클?



# 마무리

좋은 내용의 활동학습 항상 감사드립니다 🚀
내용이 조금 길지만 코다와 함께 생각해보고 싶은 내용이 많았던 것 같습니다 ..! 잘 부탁드립니다 👏

inju2403 committed:
fix: KoreanEntry에 빠진 코드 추가
description에 매칭 문자열 추가

liam-777 committed:
feat: [STEP2] repository 추가

- dummy data를 model로 바꿔주는 repository 추가
- navigation controller 스토리보드에 추가
- ListViewController 추가

liam-777 committed:
Merge branch 'k-liam.777' into feature/step2

inju2403 committed:
feat: 만국박람회 소개화면 UI 구현
스크롤뷰와 스택뷰를 사용하여 구현

liam-777 committed:
feat: [STEP2] 초기 화면 한국의 출품작 보러가기 스택 뷰 추가

- 버튼, 이미지 추가

liam-777 committed:
feat: [STEP2] 초기 화면 데이터 가져오는 로직 추가

- Reposiotry에서 Api로 네이밍 변경
- 프로퍼티 옵저버를 통해 모델 변화 감지

liam-777 committed:
perf: 데이터 fetch 작업 쓰레드 분리

liam-777 committed:
feat: [STEP2] KoreanEntryTableView 생성

- UITableViewDataSource를 채택한 Custom class 생성
- cell nib 파일 생성

liam-777 committed:
feat: [STEP2] KoreanEntryViewController 구현

- fetch data 후 프로퍼티 옵저버를 통해 tableview reload
- navigation bar 설정
- korenEntryCell 악세사리뷰 추가

liam-777 committed:
refactor: [STEP2] tableview cell 레이아웃 구조 수정

- constraint -> 스택뷰로 변경

inju2403 committed:
feat: [STEP2] KoreanEntryDetailViewContoller 구현
UITableViewDelegate를 채택한 클래스를 만들고 NotificationCenter를 이용해서 테이블 뷰 셀을 클릭했을 때 상세 화면으로 이동할 때 데이터를 넘겨주도록 구현

inju2403 committed:
refactor: [STEP2] KoreanEntryListTableViewDelegate 개선 및 코드 정리
출력 코드 제거, 공백 제거
TablewViewDelegate, TableViewDataSource를 채택하여 구현한 클래스 파일로 분리

inju2403 committed:
refactor: [STEP2] 코드 정리
폴더 이름 변경: Constants -> Common
ApiMock에서 T.self -> decodingType
ExpositionViewController에서 fetchData() 메소드 생성

liam-777 committed:
fix: [STEP2] NotificationCenter post 및 처리 메서드 수정

inju2403 committed:
chore: [STEP2] 불필요한 공백 제거

dacodaco commented on Expo1900/Expo1900/Common/Constant.swift:
```diff
@@ -0,0 +1,16 @@
+//  Created by kakao on 2022/01/11.
+//
+
+struct Constant {
+    static let koreanEntryMockFileName = "koreanEntryMock"
+    static let universalExpositionMockFileName = "universalExpositionMock"
+    static let koreanEntryListCellNibName = "KoreanEntryTableViewCell"
+    static let koreanEntryTableViewIdentifier = "ReusableCell"
+    static let showDetailKoreanEntryNotification = "showDetailKoreanEntry"
+    static let KoreanEntryDetailViewControllerIdentifier = "Detail"
+    static let selectedIndexPathRow = "selectedIndexPathRow"
+}
```
상수를 static으로 정의해 휴먼 에러를 미련에 방지하셨군요. 좋습니다 :)

dacodaco commented on Expo1900/Expo1900/Controllers/ExpositionViewController.swift:
```diff
@@ -0,0 +1,60 @@
+        }
+    }
+    
+    override func viewWillAppear(_ animated: Bool) {
+        self.navigationController?.setNavigationBarHidden(true, animated: false)
+    }
```
Life Cycle을 활용해서
목록 화면에서 다시 메인 화면으로 돌아온 경우에도 navigationBar가 숨겨질 수 있도록 했군요?

생명 주기에 대한 이해를 바탕으로 코드를 작성하신 것 같아서 뿌듯합니다. ㅎㅎ

dacodaco replied:
다만 `super.init()` 을 호출하지 않은 이유가 궁금합니다.

inju2403 replied:
실수로 빠뜨렸습니다! 추가할게요! 감사합니다ㅎㅎ

dacodaco commented on Expo1900/Expo1900/Controllers/ExpositionViewController.swift:
```diff
@@ -0,0 +1,60 @@
+        self.fetchData()
+    }
+    
+    override func viewWillDisappear(_ animated: Bool) {
```
마찬가지로 `super.init()` 을 작성하지 않은 이유가 궁금합니다.

dacodaco commented on Expo1900/Expo1900/Models/ApiMock.swift:
```diff
@@ -0,0 +1,33 @@
+            }
+        }
+    }
+}
```
Mock이라는 이름의 타입 내부에, decode 나 response 같은 데이터를 처리하는 비즈니스 로직이 담겨 있는 것이 좋을까 싶습니다.
Mock이 아닌 네트워킹을 통해 json을 디코딩해야한다면 같은 함수를 다시 만들어야할까요?

JSONService 라는 타입을 하나 만들고 관련한 로직을 담아두는 것은 어떨까요?

dacodaco commented on Expo1900/Expo1900/Models/ApiMock.swift:
```diff
@@ -0,0 +1,33 @@
+
+class ApiMock {
+    private var decoder = JSONDecoder()
+    var jsonFileName: String!
```
`jsonFileName` 에 암시적 추출 옵셔널을 사용한 이유가 궁금합니다

dacodaco replied:
+ 파일 이름을 참조할 필요가 있는게 아니라면 class 를 활용해야할 필요가 있을까요?

dacodaco commented on Expo1900/Expo1900/Controllers/ExpositionViewController.swift:
```diff
@@ -0,0 +1,60 @@
+        self.api.fetchExposition { [weak self] universalExposition in
+            self?.universalExposition = universalExposition
+        }
+    }
```
데이터의 입출력과 관련한 타입을 별도로 마련한 것은 의존성 분리 차원에서 좋은 아이디어인 것 같습니다 👍

dacodaco commented on Expo1900/Expo1900/Controllers/ExpositionViewController.swift:
```diff
@@ -0,0 +1,60 @@
+    @IBOutlet weak var durationLabel: UILabel!
+    @IBOutlet weak var descriptionLabel: UILabel!
+    
+    private var api: ExpositionApi!
```
습관적으로 ! 느낌표를 붙여서 암시적 추출 옵셔널을 활용하는 것 같아요.
개인적으로는 위험한 습관이라고 생각합니다..! 고쳐볼까요?

dacodaco commented on Expo1900/Expo1900/Controllers/ExpositionViewController.swift:
```diff
@@ -0,0 +1,60 @@
+    }
+    
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
+        let destinationViewController = segue.destination as? KoreanEntryListViewController
+        destinationViewController?.dataSource = KoreanEntryListTableViewDataSoruce()
+        destinationViewController?.delegate = KoreanEntryListTableViewDelegate()
```
저는 개인적으로 dataSource와 delegate 에 대한 의존성을 분리하고자 한게, 역으로 코드의 가독성을 해친게 아닌가 싶어요.

우리가 박람회 리스트 물품에 대한 설정을 어떻게 했는지 코드를 보고 싶을 때, 현재 상태라면 `ExpositionViewController` 에 와서야 확인할 수 있을 것 같아요. ViewController 가 자신이 가질 정보의 설정을 다른 ViewController 에게 맡긴다는 점이 조금은 이해하기 어려워요.

dacodaco replied:
아참, 만약 비슷한 TableView 를 사용하는 ViewController 가 여러 개 존재하고, 이에 대한 데이터 소스와 딜리게이트를 추상화하고 싶다면, 마냥 나쁜 방법이라고 볼 수는 없을 것 같아요. 
하지만, 지금의 기능 요구 사항에서는 비슷한 테이블 뷰를 구현할 필요가 없는 상황에서 확장성을 과하게 고려한 것은 아닌가 싶습니다..!

liam-777 replied:
코다의 말에 저희도 전적으로 동감합니다 ..!

야곰이 라이브세션때 데이터소스와 델리게이트를 분리하는 방법을 소개해주셔서 적용해보려고 시도했습니다.
그 과정에서 의존성 주입을 생각하다 보니 의존성을 최대한 낮추려는 코드를 작성하게 되었습니다.. ㅎㅎ

코다의 코멘트를 듣고, 너무 의존성에 꽂혀 있던 것 같아서 돌아보게 될 수 있었습니다. 감사합니다!!

dacodaco commented on Expo1900/Expo1900/Controllers/KoreanEntryListTableViewDelegate.swift:
```diff
@@ -0,0 +1,18 @@
+import UIKit
+
+class KoreanEntryListTableViewDelegate: NSObject, UITableViewDelegate {
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        NotificationCenter.default.post(
+            name: NSNotification.Name(Constant.showDetailKoreanEntryNotification),
+            object: nil,
+            userInfo: [Constant.selectedIndexPathRow : indexPath.row]
+        )
+    }
```
개인적으로 NotificationCenter는 변화를 수신해야하는 대상이 다수인 경우에 활용하려고 노력해요. 
지금과 같은 경우, 만약 이 딜리게이트를 활용하는 뷰가 다수라고 가정하면(재사용을 위해 추상화하셨을 테니...)
하나의 테이블 뷰를 터치했을 때,
Notification.default를 수신하고 있는 모든 테이블 뷰가 자신이 처리해야하는 알림인지를 검증해야할 것 같아요.
이런 방식의 흐름이 맞을까요?

liam-777 replied:
저희 코드 같이 tablviewDelgate과 뷰컨트롤러를 분리한 경우에는 다른 패턴을 사용한다고 해도 어느 위치에서든 분기처리를 해야될 것 같습니다.

각 테이블 뷰가 굳이 다른 동작을 필요로 하면 name으로 구별하여 보내주면 될 것 같습니다 !

dacodaco commented on Expo1900/Expo1900/Controllers/KoreanEntryListViewController.swift:
```diff
@@ -0,0 +1,76 @@
+
+    @IBOutlet weak var koreanEntryListTableView: UITableView!
+    
+    var dataSource: UITableViewDataSource!
+    var delegate: UITableViewDelegate!
```
위에서 제가 의존성을 낮추는 것에 대해서 문제점을 나열했는데,
그럼에도 불구하고 의존성을 낮추려고 노력했다는 점에서는 충분히 의의가 있다고 생각합니다.
저도 처음 보는 아이디어여서 신선했습니다.

dacodaco commented on Expo1900/Expo1900/Controllers/KoreanEntryListViewController.swift:
```diff
@@ -0,0 +1,76 @@
+    private var selectedKoreanEntry: KoreanEntry?
+    private var koreanEntryList: [KoreanEntry]? {
+        willSet(newKoreanEntryList) {
+            if let newKoreanEntryList = newKoreanEntryList,
+               let dataSource = dataSource as? KoreanEntryListTableViewDataSoruce {
+                dataSource.items = newKoreanEntryList
+                koreanEntryListTableView.reloadData()
+            }
```
API 로 부터 리스트를 받아오면,
데이터 소스가 해당 리스트를 자신의 아이템으로 옮기는 과정으로 보입니다.

뷰를 그리는데 필요한 데이터 배열이 두 공간에 배치되어 있는 것에 대해서 어떻게 생각하시나요?

dacodaco commented on Expo1900/Expo1900/Models/ExpositionApi.swift:
```diff
@@ -0,0 +1,11 @@
+//  Created by kakao on 2022/01/11.
+//
+
+protocol ExpositionApi {
+    func fetchKoreanEntryList(completion: @escaping ([KoreanEntry]?) -> Void)
+    func fetchExposition(completion: @escaping (UniversalExposition?) -> Void)
+}
```
프로토콜을 통해 의존성을 낮춘 부분 아주 좋습니다.

dacodaco commented on Expo1900/Expo1900/Models/ExpositionApiMock.swift:
```diff
@@ -0,0 +1,18 @@
+//  Created by kakao on 2022/01/11.
+//
+
+class ExpositionApiMock: ApiMock, ExpositionApi {
+    func fetchKoreanEntryList(completion: @escaping ([KoreanEntry]?) -> Void) {
+        self.jsonFileName = Constant.koreanEntryMockFileName
+        self.response(completion: completion)
+    }
+    
+    func fetchExposition(completion: @escaping (UniversalExposition?) -> Void) {
+        self.jsonFileName = Constant.universalExpositionMockFileName
+        self.response(completion: completion)
+    }
+}
```
뭔가 Mock 이라는 이름과 타입 이름이 맞지 않는 것 같아요.
이 타입은 Mock을 만드는 클래스이지 Mock 은 아니지 않을까요?

dacodaco replied:
+ 프로젝트 내부에서 대부분 클래스를 활용하고 계신 것 같은데, 지금의 경우 충분히 구조체로 대체할 수 있어보여요. 어떻게 생각하시나요?

liam-777 replied:
그래서 이름 중간에 Api를 넣었는데 이래도 어색한 네이밍일까요??

저희가 붙인 ApiMock의 이름은 더미데이터를 뜻하는 Mock과 데이터롤 fetch해오는 것을 뜻하는 api를 합쳐서 만들었는데,
코다가 생각하시는 Mock의 의미가 궁굼합니다..!

dacodaco commented on Expo1900/Expo1900/Controllers/KoreanEntryListViewController.swift:
```diff
@@ -0,0 +1,76 @@
+        )
+    }
+    
+    private func setUpNotificationObserver() {
+        NotificationCenter.default.addObserver(
+            self,
+            selector: #selector(showDetailKoreanEntry),
+            name: NSNotification.Name(Constant.showDetailKoreanEntryNotification),
+            object: nil
+        )
+    }
```
혹시 노티피케이션에 등록된 옵저버는 ViewController 가 메모리에서 해제된 이후에도 살아있을까요?

liam-777 replied:
<img width="1085" alt="스크린샷 2022-01-14 오후 1 31 03" src="https://user-images.githubusercontent.com/95397915/149451903-efbf3ca2-3353-4cb9-a235-d6a77b836ba6.png">

 https://developer.apple.com/documentation/foundation/notificationcenter/1415360-addobserver

저희가 사용한 addObserver 메서드에서는 자동으로 할당 해제해준다고 나와있습니다 !

dacodaco replied:
오오.. 좋습니닷

dacodaco commented:
파일 내부에 코멘트 남기지 못한 질문들에 대해서 제 생각을 정리해볼게요!

1. 저는 앱 전역에서 쓰이는 로직이나 서비스는 싱글턴으로 구현하는 것이 마냥 나쁘다고 생각하지 않습니다. 하지만, 그만큼 전역에서 쓰일만한 요소인지 확신할 수 있어야 겠죠? 일례로 저는 제가 만든 앱에서 사용자가 앱에 로그인했는지, 혹은 로그아웃했는지를 전달받는 서비스를 싱글턴으로 구현했습니다. 하나의 앱에서 유저가 여러명일 수 없겠다는 판단이 들어서였어요.

API 라는 별도의 파일을 만들어서 json 파일의 입력을 외부에서 관리해준 점은 참 인상적이었어요. Repository 패턴이라는 것을 한번 공부해보셔도 좋을 것 같네요 :)

2. 저는 개인적으로 스토리보드만을 View로 보기에는 무리가 있다고 생각해요. ViewController 내부에 뷰가 결국 존재하기 때문에, ViewController와 스토리보드 모두를 뷰로 보는 것이 맞다는 생각이 듭니다. 따라서 뷰컨에서 진해주고 있는 비즈니스 로직을 Controller 라는 다른 파일에서 진행해주는 것이 정통 MVC에서는 맞지 않을까... 싶습니다. 

하지만, 코코아 MVC 는 정통 MVC 와 다른 것 알고 계시나요? 저는 개인적으로 애플이 지맘대로 정해놓은 코코아 MVC 가 맘에 들지 않아요.. ㅎ 두 분은 어떤 아키텍처를 써오셨는지 궁금하네요.

4. 저도 잘 모르겠습니다...ㅎ 찾아볼게요

5. 저는 둘 다 컨트롤러에 해당한다고 생각했는데요. 어떤 이유에서 모델이라고 생각하셨는지 궁금하네요!

6. 사실 제가 xib 파일로 뷰를 구성해본 경험이 없어서 구체적인 장점을 말씀드리기는 어렵지만, 무엇보다도 기존의 방식보다 더 세밀한 요소를 다루고, 모듈화할 수 있다는 장점이 있을 것 같아요. 다른 테이블 뷰에서도 해당 셀을 원형으로 쓸 수 있다면, 충분히 유용할 것 같습니다.(물론 저는 그런 경우에 코드를 쓸 것 같긴 한데 말이죠... 회사 바이 회사니까..쩝)

7. 고민 끝에 답이 있을 겁니다. ㅎㅎ 이 부분은 제가 먼저 말씀 드리지는 않을게요.

liam-777 committed:
refactor: [STEP2] PR 리뷰 반영

- Mock API 네이밍 수정 및 기능 분리 -> JSONFileDecodable
- class 상속을 제거하고 POP 방식으로 수정
- 외부 생성을 막기 위해 ExpositionRepositoryInjection에 private 생성자 추가
- 암시적 추출 옵셔널 제거
- super.init() 추가

liam-777 commented:
@dacodaco 

PR 리뷰 내용 감사합니다 👍
수정해서 커밋 반영하였습니다 ! 시간나실 때 머지 부탁드리겠습니다 !!
